Implement tracking architecture

gbirke opened the pull request:
Entry points pass mapping factories to Tracker, Tracker classes convert Domain events to specific tracker events. The list of allowed events determines which events are actually sent.

We'll have to see how we can support sending the dynamic banner size to the tracker, maybe we'll refactor that.

gbirke committed:
Start tracking and refactor form controllers

- Add tests for step controllers
- Implement new FormStep controllers for other channels
    - implements usage of form step controllers for desktop, english, mobile (DE)
- Further refactoring of StepControllers
    - Use slot names to create an index map
    - Make Parameter names for steps more generic
    - Define Naming Schema for steps:
      create[Submittable|Intermediate][ComponentName][PossibleQualifier]
- Implement and refactor StepControllers
    - We now use the `goToStep` function exclusively to explicitly wire up
      each step controller with all the possible pages it can go to.
    - Using page indexes also allows us to keep the form components and their
      step controllers very closely together in the setup function of the
      top-level components. Hovering over an index highlights all the usages
      in the IDE.
    - Only desktop DE has the new step controllers.
- Get rid of page controllers, put them into step controllers instead
- Get rid of Next event in form controllers
    - we use the submit or previous event to handle things happening
      when switching to a different form
    - introduces "isCurrent" as an indicator for whether a form page is
      currently visible for the user. This helps with unique tracking calls.
- Add decrease and increase event to all channels
- Add increase and decrease custom amount events
- Add CustomAmountFormPageShownEvent
- Add UpgradeToYearlyFormPageShownEvent tracking
    - Rename FormControllerCtrl to FormController where banners share a
      FormController implementation.
    - Add tests for FormControllers
- Add Expand mini banner event and start with already donated event
- Fix CI for WPDE banners
- Start implementing tracking architecture
    - Entry points pass mapping factories to Tracker, Tracker classes convert
      Domain events to specific tracker events. The list of allowed events
      determines which events are actually sent.
    - We'll have to see how we can support sending the dynamic banner size to
      the tracker, maybe we'll refactor that.

Co-Authored-By: Gabriel Birke <gabriel.birke@wikimedia.de>
Co-Authored-By: Corinna Hillebrand <corinna.hillebrand@wikimedia.de>
Co-Authored-By: Sperling-0 <tanuja.doriya@wikimedia.de>
Co-Authored-By: Abban Dunne <abban.dunne@wikimedia.de>

Sperling-0 committed:
Implement event tracking

Co-authoured by: Gabriel Birke <Gabriel.birke@wikimedia.de>
Co-authoured by: Corinna Hillebrand <Corinna.hillebrand@wikimedia.de>
Please enter the commit message for your changes. Lines starting

Abban committed:
Prepare code for new tracking schema (#99)

This modifies the interfaces to allow easier implementation
of the new tracking schema when we get to it.

* Moved event rate out of events and into maps
* Added event data interface based off the new schema description
* Marked old interfaces and implementations as deprecated
* Fixed tests and linting issues

gbirke committed:
Add more Event handling

Extract useFormStepShownEvent for all form pages

Co-Authored-By: Corinna Hillebrand <corinna.hillebrand@wikimedia.de>
Co-Authored-By: Sperling-0 <tanuja.doriya@wikimedia.de>

Abban committed:
Fix MultistepDonationForm tests

* Fixes the tests in the MultiStepDonationForm
* Fix the sub form tests
* Remove FormController interfaces and replace emit data

gbirke committed:
Add remaining trackings

Add correct close event tracking
Add namespace-tracking
Add already-donated tracking

Add missing current-page tracking to button forms
Add UpgradeToYearlyButtonForm tracking tests

Co-Authored-By: Corinna Hillebrand <corinna.hillebrand@wikimedia.de>
Co-Authored-By: Sperling-0 <tanuja.doriya@wikimedia.de>

gbirke committed:
Start with TrackerWPDE

Copied code from old banners
Test still needs to be ported & adapted

Co-Authored-By: Corinna Hillebrand <corinna.hillebrand@wikimedia.de>
Co-Authored-By: Sperling-0 <tanuja.doriya@wikimedia.de>

Abban committed:
Make tests pass

Abban committed:
Merge pull request #101 from wmde/multistep-form-tests

Fix MultistepDonationForm tests

Abban committed:
Add locale to English banners

Abban committed:
Make MainDonationForm button slottable

Abban committed:
Add default slots to keen slider

Abban committed:
Fix unused imports

Abban committed:
Remove component styles from WPDE mobile

Abban committed:
Add .prevent to all posible anchor link events

Abban commented on src/tracking/TrackerWPDE.ts:
```diff
@@ -0,0 +1,108 @@
+	 * @param {TrackingEvent} event
+	 * @private
+	 */
+	private getEventNameFromEvent( event: TrackingEvent ): string {
```
I'd be inclined to add a property to the events for this. It could be implemented as a get property that does the concatenation

gbirke replied:
But then all the events in WPOrg banners would carry around that unneccessary code, which is why we decided to stick that switch statement in here, where it only affects 2 banners, which need this anyway.

Abban committed:
Add tests for WPDETracker

gbirke committed:
Finish WPDE tracker

Add more tests for TrackerWPDE
Use tracking rate in TrackerWPDE

Also add tests for LegacyTrackerWPORG

Co-Authored-By: Gabriel Birke <gabriel.birke@wikimedia.de>
Co-Authored-By: Corinna Hillebrand <corinna.hillebrand@wikimedia.de>
Co-Authored-By: Sperling-0 <tanuja.doriya@wikimedia.de>

gbirke committed:
Change SizeIssue tracking

With the new event schema we can no longer rely on the event adding the
sizes for certain events. Instead, we generate the viewport data where
it's needed (only for NotShownState.SizeIssue) and pass it to the event as custom data.

Co-Authored-By: Corinna Hillebrand <corinna.hillebrand@wikimedia.de>
Co-Authored-By: Sperling-0 <tanuja.doriya@wikimedia.de>
Co-Authored-By: Gabriel Birke <gabriel.birke@wikimedia.de>

gbirke committed:
Change close event tracking

Renamed userInteraction to userChoice because that words makes it more
clear that it's a choice between several options and not the name of an
interaction.

Have a unified close event that doesn't use CloseSources, but translates
them to feature and userChoice pairs. The switch statement is
unfortunate, maybe we can find a better mechanism.

Adapt legacy event maps

Co-Authored-By: Gabriel Birke <gabriel.birke@wikimedia.de>
Co-Authored-By: Corinna Hillebrand <corinna.hillebrand@wikimedia.de>
Co-Authored-By: Sperling-0 <tanuja.doriya@wikimedia.de>

Abban committed:
Refactor close tracking event

The close events are now created in the components
and passed up the tree.

The CloseSources enum is renamed to LegacyCloseSources
and marked as deprecated.

The long switch statement is removed from the close event
and the conversion is done in the legacy closeMapper.

Updated the WPDE banner mappings to track a single close
event.

Abban committed:
Convert not shown event for new schema

Rename the not shown events, and add a mapper that converts the
new names to the old ones for the legacy schema.

Abban committed:
Tidy up project

Move some interfaces into Domain, and delete unused

Abban committed:
Add feature tests for multistep forms

* Add tests for the different form flows
* Removes form controller references from banner tests
* Update desktop content feature to split test signatures
* Add workaround to vue test-utils for catching form submits
  in sub-components

moiikana commented:
I'd like to propose this additional change here:
https://github.com/wmde/fundraising-banners/pull/103 (outdated)
https://github.com/wmde/fundraising-banners/pull/109 (<-- using a union type)

moiikana committed:
Add Union Type for more control over feature name strings

moiikana committed:
Change types to new Union Type

- use random values from the union type in the tests

Abban committed:
Merge pull request #102 from wmde/form-feature-tests

Add feature tests for multistep forms

Abban committed:
Merge pull request #109 from wmde/use-union-type

Use union type
